HPX 1.4.0rc1 does not link on s390x

## Expected Behavior

HPX should compile on s390x as hpx 1.3.0 compiled there.

## Actual Behavior

Linking error for 

```
.a ../lib/libhpx_resource_partitioner.a ../lib/libhpx_cache.a ../lib/libhpx_coroutines.a ../lib/libhpx_util.a ../lib/libhpx_memory.a ../lib/libhpx_hardware.a ../lib/libhpx_plugin.a ../lib/libhpx_program_options.a /usr/lib64/libboost_program_options.so ../lib/libhpx_algorithms.a ../lib/libhpx_statistics.a ../lib/libhpx_execution.a ../lib/libhpx_allocator_support.a /usr/lib64/libtcmalloc_minimal.so ../lib/libhpx_topology.a /usr/lib64/libhwloc.so ../lib/libhpx_iterator_support.a ../lib/libhpx_functional.a ../lib/libhpx_concurrency.a ../lib/libhpx_basic_execution.a ../lib/libhpx_serialization.a ../lib/libhpx_debugging.a ../lib/libhpx_errors.a ../lib/libhpx_datastructures.a ../lib/libhpx_format.a ../lib/libhpx_logging.a ../lib/libhpx_filesystem.a ../lib/libhpx_timing.a ../lib/libhpx_thread_support.a ../lib/libhpx_hashing.a ../lib/libhpx_concepts.a ../lib/libhpx_type_support.a ../lib/libhpx_concepts.a ../lib/libhpx_type_support.a ../lib/libhpx_assertion.a ../lib/libhpx_config.a /usr/lib64/libboost_system.so -lpthread ../lib/libhpx_preprocessor.a 
/usr/bin/ld: CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: in function `hpx::threads::coroutines::detail::lx::swap_context(hpx::threads::coroutines::detail::lx::x86_linux_context_impl_base&, hpx::threads::coroutines::detail::lx::x86_linux_context_impl_base const&, hpx::threads::coroutines::detail::default_hint)':
/builddir/build/BUILD/hpx_1.4.0/libs/coroutines/include/hpx/coroutines/detail/context_linux_x86.hpp:531: undefined reference to `swapcontext_stack'
/usr/bin/ld: CMakeFiles/hpx.dir/runtime/threads/execution_agent.cpp.o: in function `hpx::threads::coroutines::detail::lx::swap_context(hpx::threads::coroutines::detail::lx::x86_linux_context_impl_base&, hpx::threads::coroutines::detail::lx::x86_linux_context_impl_base const&, hpx::threads::coroutines::detail::yield_hint)':
/builddir/build/BUILD/hpx_1.4.0/libs/coroutines/include/hpx/coroutines/detail/context_linux_x86.hpp:542: undefined reference to `swapcontext_stack'
/usr/bin/ld: /builddir/build/BUILD/hpx_1.4.0/libs/coroutines/include/hpx/coroutines/detail/context_linux_x86.hpp:542: undefined reference to `swapcontext_stack'
/usr/bin/ld: ../lib/libhpx_coroutines.a(context_base.cpp.o): in function `hpx::threads::coroutines::detail::lx::swap_context(hpx::threads::coroutines::detail::lx::x86_linux_context_impl_base&, hpx::threads::coroutines::detail::lx::x86_linux_context_impl_base const&, hpx::threads::coroutines::detail::default_hint)':
/builddir/build/BUILD/hpx_1.4.0/libs/coroutines/include/hpx/coroutines/detail/context_linux_x86.hpp:531: undefined reference to `swapcontext_stack'
/usr/bin/ld: /builddir/build/BUILD/hpx_1.4.0/libs/coroutines/include/hpx/coroutines/detail/context_linux_x86.hpp:531: undefined reference to `swapcontext_stack'
/usr/bin/ld: ../lib/libhpx_coroutines.a(context_base.cpp.o):/builddir/build/BUILD/hpx_1.4.0/libs/coroutines/include/hpx/coroutines/detail/context_linux_x86.hpp:531: more undefined references to `swapcontext_stack' follow
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/hpx.dir/build.make:2478: lib/libhpx.so.1.4.0] Error 1
make[2]: Leaving directory '/builddir/build/BUILD/hpx_1.4.0/serial'
make[1]: *** [CMakeFiles/Makefile2:40314: src/CMakeFiles/hpx.dir/all] Error 2
make: *** [Makefile:144: all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.pqNZR6 (%build)
RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.pqNZR6 (%build)
Child return code was: 1
EXCEPTION: [Error()]
```


## Steps to Reproduce the Problem

Build HPX with on s390x, please look at the error log on [Fedora's build system](https://koji.fedoraproject.org/koji/getfile?taskID=38997711&volume=DEFAULT&name=build.log&offset=-4000) for more details.

## Specifications

... Please describe your environment

  - HPX Version: 1.4.1-rc1
  - Platform (compiler, OS): Fedora rawhide

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1171,6 +1171,20 @@ if(HPX_WITH_CXX14_RETURN_TYPE_DEDUCTION)
 endif()
 
 ################################################################################
+# Set configuration option to use Boost.Context or not. This depends on the
+# platform.
+set(__use_generic_coroutine_context OFF)
+if(APPLE)
+  set(__use_generic_coroutine_context ON)
+endif()
+if("${HPX_PLATFORM_UC}" STREQUAL "BLUEGENEQ")
+  set(__use_generic_coroutine_context ON)
+endif()
+hpx_option(HPX_WITH_GENERIC_CONTEXT_COROUTINES BOOL
+  "Use Boost.Context as the underlying coroutines context switch implementation."
+  ${__use_generic_coroutine_context} ADVANCED)
+
+################################################################################
 # check for miscellaneous things
 ################################################################################
 
@@ -1561,20 +1575,6 @@ endif()
 
 # store target architecture for later use
 set(HPX_WITH_TARGET_ARCHITECTURE ${__target_arch} CACHE INTERNAL "" FORCE)
-
-################################################################################
-# Set configuration option to use Boost.Context or not. This depends on the
-# platform.
-set(__use_generic_coroutine_context OFF)
-if(APPLE)
-  set(__use_generic_coroutine_context ON)
-endif()
-if(HPX_PLATFORM_UC STREQUAL "BLUEGENEQ")
-  set(__use_generic_coroutine_context ON)
-endif()
-hpx_option(HPX_WITH_GENERIC_CONTEXT_COROUTINES BOOL
-  "Use Boost.Context as the underlying coroutines context switch implementation."
-  ${__use_generic_coroutine_context} ADVANCED)
 
 # Compatibility with using Boost.ProgramOptions, introduced in V1.4.0
 hpx_option(HPX_PROGRAM_OPTIONS_WITH_BOOST_PROGRAM_OPTIONS_COMPATIBILITY

--- a/libs/coroutines/include/hpx/coroutines/detail/context_impl.hpp
+++ b/libs/coroutines/include/hpx/coroutines/detail/context_impl.hpp
@@ -116,7 +116,7 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
 }}}}    // namespace hpx::threads::coroutines::detail
 
 #elif (defined(__linux) || defined(linux) || defined(__linux__)) &&            \
-    !defined(__bgq__) && !defined(__powerpc__) && !defined(__s390x_)
+    !defined(__bgq__) && !defined(__powerpc__) && !defined(__s390x__)
 
 #include <hpx/coroutines/detail/context_linux_x86.hpp>
 namespace hpx { namespace threads { namespace coroutines { namespace detail {
@@ -125,7 +125,7 @@ namespace hpx { namespace threads { namespace coroutines { namespace detail {
 }}}}    // namespace hpx::threads::coroutines::detail
 
 #elif defined(_POSIX_VERSION) || defined(__bgq__) || defined(__powerpc__) ||   \
-    defined(__s390x_)
+    defined(__s390x__)
 
 #include <hpx/coroutines/detail/context_posix.hpp>
 namespace hpx { namespace threads { namespace coroutines { namespace detail {

--- a/libs/hashing/include/hpx/hashing/fibhash.hpp
+++ b/libs/hashing/include/hpx/hashing/fibhash.hpp
@@ -12,14 +12,14 @@
 
 #include <hpx/config.hpp>
 
-#include <cstddef>
+#include <cstdint>
 #include <cstdlib>
 
 namespace hpx { namespace util {
 
     namespace detail {
 
-        template <std::size_t N>
+        template <std::uint64_t N>
         struct hash_helper;
 
         template <>
@@ -28,7 +28,7 @@ namespace hpx { namespace util {
             HPX_STATIC_CONSTEXPR int log2 = -1;
         };
 
-        template <std::size_t N>
+        template <std::uint64_t N>
         struct hash_helper
         {
             HPX_STATIC_CONSTEXPR std::uint64_t log2 =
@@ -36,13 +36,14 @@ namespace hpx { namespace util {
             HPX_STATIC_CONSTEXPR std::uint64_t shift_amount = 64 - log2;
         };
 
-        HPX_STATIC_CONSTEXPR std::uint64_t golden_ratio = 11400714819323198485llu;
+        HPX_STATIC_CONSTEXPR std::uint64_t golden_ratio =
+            11400714819323198485llu;
     }    // namespace detail
 
     // This function calculates the hash based on a multiplicative Fibonacci
     // scheme
-    template <std::size_t N>
-    HPX_CONSTEXPR std::size_t fibhash(std::uint64_t i)
+    template <std::uint64_t N>
+    HPX_CONSTEXPR std::uint64_t fibhash(std::uint64_t i)
     {
         using helper = detail::hash_helper<N>;
         static_assert(N != 0, "This algorithm only works with N != 0");

--- a/libs/hashing/include/hpx/hashing/fibhash.hpp
+++ b/libs/hashing/include/hpx/hashing/fibhash.hpp
@@ -16,7 +16,9 @@
 #include <cstdlib>
 
 namespace hpx { namespace util {
+
     namespace detail {
+
         template <std::size_t N>
         struct hash_helper;
 
@@ -29,18 +31,18 @@ namespace hpx { namespace util {
         template <std::size_t N>
         struct hash_helper
         {
-            HPX_STATIC_CONSTEXPR std::size_t log2 =
+            HPX_STATIC_CONSTEXPR std::uint64_t log2 =
                 hash_helper<(N >> 1)>::log2 + 1;
-            HPX_STATIC_CONSTEXPR std::size_t shift_amount = 64 - log2;
+            HPX_STATIC_CONSTEXPR std::uint64_t shift_amount = 64 - log2;
         };
 
-        HPX_STATIC_CONSTEXPR std::size_t golden_ratio = 11400714819323198485llu;
+        HPX_STATIC_CONSTEXPR std::uint64_t golden_ratio = 11400714819323198485llu;
     }    // namespace detail
 
     // This function calculates the hash based on a multiplicative Fibonacci
     // scheme
     template <std::size_t N>
-    HPX_CONSTEXPR std::size_t fibhash(std::size_t i)
+    HPX_CONSTEXPR std::size_t fibhash(std::uint64_t i)
     {
         using helper = detail::hash_helper<N>;
         static_assert(N != 0, "This algorithm only works with N != 0");

